asv_bench/benchmarks/omnisci/benchmarks.py

```diff
@@ -28,21 +28,60 @@
     trigger_import,
     get_benchmark_shapes,
 )
+from ..utils.common import random_state
+import numpy as np
 
 
 class TimeJoin:
-    param_names = ["shape", "how"]
+    param_names = ["shape", "how", "is_equal_keys"]
     params = [
         get_benchmark_shapes("omnisci.TimeJoin"),
         ["left", "inner"],
+        [True, False],
     ]
 
-    def setup(self, shape, how):
-        self.df1 = generate_dataframe(ASV_USE_IMPL, "int", *shape, RAND_LOW, RAND_HIGH)
-        self.df2 = generate_dataframe(ASV_USE_IMPL, "int", *shape, RAND_LOW, RAND_HIGH)
+    def setup(self, shape, how, is_equal_keys):
+        self.df1, self.df2 = (
+            generate_dataframe(
+                ASV_USE_IMPL,
+                "int",
+                *frame_shape,
+                RAND_LOW,
+                RAND_HIGH,
+                cache_prefix=f"{i}-th_frame_to_join",
+            )
+            for i, frame_shape in enumerate((shape, shape))
+        )
+
+        if is_equal_keys:
+            # When the frames have default indices to join on: RangeIndex(frame_length),
+            # OmniSci backend performs join on the internal meta-column called 'rowid'.
+            # There is a bug in the engine that makes such joins fail. To avoid joining
+            # on the meta-column we explicitly specify a non-default index to join on.
+            # https://github.com/modin-project/modin/issues/3442
+            common_index = np.arange(1, len(self.df1) + 1)
+            self.df1.index = common_index
+            self.df2.index = common_index
```

Comment:
```suggestion
            self.df1.index = self.df2.index = common_index
```

Reply:
Rewritten it into the following, to avoid possibilities of using one shared index object:
```python
self.df1.index = np.arange(...)
self.df2.index = np.arange(...)
```

Probably this is an excessive safety because since the input object for the index are numpy arrays, we are forced to convert them to pandas Indices and so create a new object every time we assign a new index:
```python
>>> common_index_np = np.arange(1, len(df1) + 1)
>>> df1.index = common_index_np
>>> df2.index = common_index_np
>>> df1.index is df2.index
False
```
But if someday somebody would change a numpy array to pandas Index, then we will get a shared index objects, which we don't want to:
```python
>>> common_index_pandas = pandas.RangeIndex(1, len(df1) + 1)
>>> df1.index = common_index_pandas
>>> df2.index = common_index_pandas
>>> df1.index is df2.index
True
```
So I just create a new index object for every frame with the appropriate in-code comment to avoid this.

Reply:
Good point!

So you can just add `assert not df1.index is df2.index` with comment that we don't want use shared object here. Right?

Reply:
I don't think we need an `assert` here, it's really excessive since we explicitly generate new objects for every index + there is a comment explaining why it shouldn't be changed.

Reply:
I mean assert check instead of generate new objects for every index. So we will save time and explicitly say what the problem may be.

Reply:
I don't like the idea of putting an assert to the benchmark, because if one day the following code would return `True` we will have to rewrite our benchmarks and manually restore measurements history 

```
>>> common_index_np = np.arange(1, len(df1) + 1)
>>> df1.index = common_index_np
>>> df2.index = common_index_np
>>> df1.index is df2.index
False
```

Reply:
Can't this only happen when we change the code anyway? :)

Reply:
> So we will save time

Actually, creating a new index via `pandas.RangeIndex` is much more efficient than via `np.arange`, since RangeIndex is purely a range, stores only its metadata.

So, I just got rid of using `np.arange` and replaced it with `pandas.RangeIndex`. This way, we <i>must</i> generate a new index object for every frame (see [this](https://github.com/modin-project/modin/pull/3721#discussion_r755197850)), but do it in the most efficient way.

Reply:
ok